File: GE-Grader-Examination/G6702-Exam-2567-S2/Grader-03/2567_2_Q3_C1/2567_2_Q3_C1.cpp
#include<iostream>
#include<vector>
using namespace std;

int N, M;
bool is_game_end = false;
char PLAYER[2] = {'1', '2'};
int DIRECTIONS[8][2] = {{0, 1}, {0, -1}, {-1, 0}, {1, 0}, {-1, 1}, {1, -1}, {-1, -1}, {1, 1}};
vector<int> column_sequence, column_pieces;
vector<vector<char>> board;

// Get row index of a piece
int get_row(const int &col)
{
    return M - (column_pieces[col] + 1);
}

// Check if the (row, col) is inside the board
bool is_inside_board(const int &row, const int &col)
{
    return (row >= 0) && (row < M) && (col >= 0) && (col < N);
}

// Check if a piece in (row, col) belongs to a player
bool is_player_piece(const int &row, const int &col, const char &player)
{
    return is_inside_board(row, col) && board[row][col] == player;
}

// Display board
void display_board()
{
    for(vector<char> &row : board)
    {
        for(char &c : row) { cout << c; }
        cout << '\n';
    }
}

// A player drop a piece into a board
bool drop_piece(const int &row, const int &col, const char &player)
{
    if(is_inside_board(row, col))
    {
        board[row][col] = player;
        ++column_pieces[col];
        return true;
    }
    return false;
}

// Check if the player wins a game
// DIRECTIONS
// - East      (0, 1)   | West      (0, -1)
// - North     (-1, 0)  | South     (1, 0)
// - Northeast (-1, 1)  | Southwest (1, -1)
// - Northwest (-1, -1) | Southeast (1, 1)
bool is_player_win(const int &row, const int &col, const char &player)
{
    // Count 4 consecutive pieces in each direction
    int count = 1;
    for(int i = 0; i < 8; i++)
    {
        // Reset when count on a new direction
        if(i % 2 == 0) { count = 1; }
        // Counting
        int dr = DIRECTIONS[i][0];
        int dc = DIRECTIONS[i][1];
        for(int x = 1; x <= 3; x++)
        {
            if(!is_player_piece(row + (dr * x), col + (dc * x), player)) { break; }
            ++count;
        }
        // The player has 4 consecutive pieces
        if(count >= 4) { return true; }
    }
    return false;
}

int main()
{
    // This makes std::cin and std::cout faster
    ios_base::sync_with_stdio(false); cin.tie(0);

    // Input and construct board
    cin >> N >> M;
    int temp;
    while(cin >> temp)
    {
        column_sequence.push_back(temp - 1);
    }
    board.assign(M, vector<char>(N, '.'));
    column_pieces.assign(N, 0);

    // Play a game
    for(int turn = 0; turn < column_sequence.size(); turn++)
    {
        // Get player information
        int col = column_sequence[turn];
        int row = get_row(col);
        char player = PLAYER[turn % 2];
        // Drop a piece
        bool is_column_full = !(drop_piece(row, col, player));

        // Check if the player wins
        if(turn >= 3 && is_player_win(row, col, player))
        {
            cout << "Player " << player << " wins\n";
            is_game_end = true;
            break;
        }
        // Check if the column is full
        if(turn >= M - 1 && is_column_full)
        {
            cout << "Column full\n";
            is_game_end = true;
            break;
        }
        // Check if the board is full
        if(turn == (M * N) - 1)
        {
            cout << "Board full\n";
            is_game_end = true;
            break;
        }
    }
    // No one wins after all turns
    if(!is_game_end) { cout << "No more moves\n"; }
    display_board();
    return 0;
}

File: GE-Grader-Examination/G6702-Exam-2567-S2/Grader-03/2567_2_Q3_C1/2567_2_Q3_C1.py
# --------------------------------------------------
# File Name : 2567_2_Q3_C1.py
# Problem   : Connect Four
# Author    : Worralop Srichainont
# Date      : 2025-07-30
# --------------------------------------------------

# Constants for directions
# East      ( 0,  1) | West      (0, -1)
# North     (-1,  0) | South     (1,  0)
# Northeast (-1,  1) | Southwest (1, -1)
# Northwest (-1, -1) | Southeast (1,  1)
DIRECTIONS = ((0, 1), (0, -1), (-1, 0), (1, 0), (-1, 1), (1, -1), (-1, -1), (1, 1))

# Game Configuration
PLAYER = ["1", "2"]
WINNING_PIECE_COUNT = 4

# Input amount of rows and columns of the board
cols, rows = [int(num) for num in input().split()]

# Initialize the game board and pieces on each column
board = [["."] * cols for _ in range(rows)]
pieces_on_column = [0] * cols

# Input the order of dropping pieces to each columns
drop_on_column_order = [int(num) - 1 for num in input().split()]


# Check if the given row and column are inside the board
def is_inside_board(row, col):
    return (0 <= row < rows) and (0 <= col < cols)


# Calculate the row to drop a piece in the given column
def get_drop_piece_row(col):
    return rows - (pieces_on_column[col] + 1)


# Check if the piece at the given row and column is the player's piece
def is_player_piece(row, col, player):
    return is_inside_board(row, col) and board[row][col] == player


# Drop a piece in the given row and column for the player
# Return True if the piece is dropped successfully, otherwise return False
def drop_piece(row, col, player):
    if is_inside_board(row, col):
        board[row][col] = player
        pieces_on_column[col] += 1
        return True
    return False


# Check if the player has won by connecting 4 pieces in a row
def is_player_win(row, col, player):
    # Initialize the piece count
    piece_count = 1

    # Check all 8 directions for a win
    for i in range(len(DIRECTIONS)):
        # Get the current direction
        dr, dc = DIRECTIONS[i]
        # Reset piece count when resetting direction
        if i % 2 == 0:
            piece_count = 1

        # Count pieces in the current direction
        for n in range(1, WINNING_PIECE_COUNT):
            if not is_player_piece(row + (dr * n), col + (dc * n), player):
                break
            piece_count += 1

        # Check if the player has connected enough pieces
        if piece_count >= WINNING_PIECE_COUNT:
            return True
    # If no direction has enough pieces, return False
    return False


# Display the game board
def display_board():
    for row in board:
        print("".join(row))


# Main function to play the game
def play_game():
    # Initialize the game state
    is_game_end = False

    # Process each turn in the order of dropping pieces
    for turn in range(len(drop_on_column_order)):
        # Get the row and column to drop the piece
        col = drop_on_column_order[turn]
        row = get_drop_piece_row(col)
        player = PLAYER[turn % 2]

        # Player drops a piece in the specified column
        # If the column is full, the piece cannot be dropped
        is_column_full = not drop_piece(row, col, player)

        # If the player has won, display the board and end the game
        if turn >= 3 and is_player_win(row, col, player):
            print("Player", player, "wins")
            display_board()
            is_game_end = True
            break

        # If the column is full, display a message and end the game
        if turn >= rows - 1 and is_column_full:
            print("Column full")
            display_board()
            is_game_end = True
            break

        # If the board is full, display a message and end the game
        if turn == (rows * cols) - 1:
            print("Board full")
            display_board()
            is_game_end = True
            break

    # If the game has not ended after all turns, display a message
    if not is_game_end:
        print("No more moves")
        display_board()


# Run the function to play the game
play_game()
